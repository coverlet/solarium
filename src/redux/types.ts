export interface IRedux {
  app: IApp;
  validators: IValidators;
}

export interface IApp {
  cluster: string;
  count: number;
}

export interface IValidatorInfo {
  network: string;
  account: string;
  name: string;
  keybase_id: string;
  www_url: string;
  details: string;
  created_at: string;
  updated_at: string;
  total_score: number;
  root_distance_score: number;
  vote_distance_score: number;
  skipped_slot_score: number;
  software_version: string;
  software_version_score: number;
  stake_concentration_score: number;
  data_center_concentration_score: number;
  published_information_score: number;
  security_report_score: number;
  active_stake: string;
  active_stake_sol?: number;
  commission: number;
  delinquent: string;
  vote_account: string;
  skipped_slots: number;
  skipped_slot_percent: string;
  ping_time: number;
  url: string;
  pic?: {
    file: string;
    x: number;
    y: number;
  };
}

export enum EValidatorsSort {
  fee,
  stake,
  score,
}

export interface IValidatorsSorter {
  by: EValidatorsSort;
  direction: 'asc' | 'desc';
}

export interface IValidators {
  fetchingCluster: string;
  isFetching: boolean;
  validators: IValidatorInfo[];
  display: {
    sort: IValidatorsSorter;
  };
}

// TODO normalize validators to this type
export interface IValidatorsList {
  [key: string]: IValidatorInfo;
}
