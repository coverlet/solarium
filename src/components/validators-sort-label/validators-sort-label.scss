a.validators-sort-label {
  text-decoration: none;
  color: #fff;

  .rs-icon {
    font-size: 11px;
    font-weight: bold;
    padding-left: 4px;
  }
}
