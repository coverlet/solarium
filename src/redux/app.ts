import { createSlice } from '@reduxjs/toolkit';
import { defaultCluster } from '../config/clusters';
import { IApp, IRedux } from './types';

const initialState: IApp = {
  cluster: defaultCluster,
  count: 0,
};

const appSlice = createSlice({
  name: 'app',
  initialState,
  reducers: {
    setCluster: (state, action) => ({
      ...state,
      cluster: action.payload,
    }),
    setCount: (state, action) => ({
      ...state,
      count: action.payload,
    }),
  },
});

export const { setCount, setCluster } = appSlice.actions;

export const selectCount = (state: IRedux): number => state.app.count;
export const selectCluster = (state: IRedux): string => state.app.cluster;

export default appSlice.reducer;
