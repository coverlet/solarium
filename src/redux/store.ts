import { configureStore } from '@reduxjs/toolkit';
import appReducer from './app';

// TODO MAYBE REMOVE SS Redux
// TODO check this out

export default configureStore({
  reducer: {
    app: appReducer,
  },
  devTools: process.env.NODE_ENV === 'development',
});
