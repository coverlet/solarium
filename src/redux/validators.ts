import { createSlice, Dispatch } from '@reduxjs/toolkit';
import { getValidators } from '../client/api';
import { lamportsToSol } from '../utils/convert-sol';
import { IRedux, IValidatorInfo, IValidators, ValidatorsSort } from './types';

const initialState: IValidators = {
  fetchingCluster: '',
  isFetching: false,
  validators: [],
  display: {
    sort: {
      by: ValidatorsSort.stake,
      direction: 'desc',
    },
  },
};

const validatorsSlice = createSlice({
  name: 'validators',
  initialState,
  reducers: {
    setValidators: (state, action) => ({
      ...state,
      fetchingCluster: '',
      isFetching: false,
      validators: action.payload,
    }),
    setFetching: (state, action) => ({
      ...state,
      fetchingCluster: action.payload.cluster,
      isFetching: action.payload.isFetching,
    }),
  },
});

export const { setValidators, setFetching } = validatorsSlice.actions;

export const selectValidators = (state: IRedux): IValidatorInfo[] => state.validators.validators;
export const selectValidatorsLoading = (state: IRedux): boolean => state.validators.isFetching;

export default validatorsSlice.reducer;

export const fetchValidators = () => (dispatch: Dispatch, getState: () => IRedux): void => {
  const state: IRedux = getState();

  // already fetching validators for this cluster
  if (state.validators.isFetching && state.validators.fetchingCluster === state.app.cluster) {
    return;
  }

  dispatch(setFetching({ cluster: state.app.cluster, isFetching: true }));

  getValidators(state.app.cluster).then((data) => {
    // XXX is this really that bad?...
    // populate active_stake_sol
    data.forEach((v: IValidatorInfo) => {
      v.active_stake_sol = lamportsToSol(v.active_stake);
    });
    dispatch(setValidators(data));
  });
};
