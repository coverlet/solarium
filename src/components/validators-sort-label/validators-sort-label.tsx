/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { ReactElement } from 'react';
import { Icon } from 'rsuite';
import { EValidatorsSort, IValidatorsSorter } from '../../redux/types';

import './validators-sort-label.scss';

export interface IProps {
  sort: IValidatorsSorter;
  label: string;
  onClick: () => void;
  sorter: EValidatorsSort;
}

export const ValidatorsSortLabel = ({ sort, label, onClick, sorter }: IProps): ReactElement => {
  return (
    <a
      className="validators-sort-label"
      href="#"
      onClick={(e) => {
        e.preventDefault();
        onClick();
      }}
    >
      {label}
      {sort.by === sorter && (
        <Icon icon={sort.direction === 'asc' ? 'long-arrow-up' : 'long-arrow-down'} />
      )}
    </a>
  );
};
