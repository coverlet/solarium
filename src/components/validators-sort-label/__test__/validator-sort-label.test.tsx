import React from 'react';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { EValidatorsSort } from '../../../redux/types';
import { ValidatorsSortLabel, IProps } from '../validators-sort-label';

const clickMock = jest.fn();

const props: IProps = {
  label: 'LABEL',
  onClick: clickMock,
  sort: {
    by: EValidatorsSort.fee,
    direction: 'asc',
  },
  sorter: EValidatorsSort.score,
};

afterEach(cleanup);

describe('footer components', () => {
  it('renders', () => {
    const { asFragment } = render(<ValidatorsSortLabel {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays icon', () => {
    let rendered = render(<ValidatorsSortLabel {...props} sorter={EValidatorsSort.fee} />);
    expect(rendered.asFragment()).toMatchSnapshot();
    rendered = render(
      <ValidatorsSortLabel
        {...props}
        sorter={EValidatorsSort.fee}
        sort={{ ...props.sort, direction: 'desc' }}
      />
    );
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  it('calls on click function', () => {
    render(<ValidatorsSortLabel {...props} />);
    const button = screen.getByRole('link', { name: /label/i });
    fireEvent.click(button);

    expect(clickMock).toBeCalled();
  });
});
