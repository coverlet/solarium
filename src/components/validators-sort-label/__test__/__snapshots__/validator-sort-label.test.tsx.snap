// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`footer components displays icon 1`] = `
<DocumentFragment>
  <a
    class="validators-sort-label"
    href="#"
  >
    LABEL
    <i
      class="rs-icon rs-icon-long-arrow-up"
    />
  </a>
</DocumentFragment>
`;

exports[`footer components displays icon 2`] = `
<DocumentFragment>
  <a
    class="validators-sort-label"
    href="#"
  >
    LABEL
    <i
      class="rs-icon rs-icon-long-arrow-down"
    />
  </a>
</DocumentFragment>
`;

exports[`footer components renders 1`] = `
<DocumentFragment>
  <a
    class="validators-sort-label"
    href="#"
  >
    LABEL
  </a>
</DocumentFragment>
`;
